apis/metal3.io/v1alpha1/firmwareschema_types.go

```diff
@@ -0,0 +1,163 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"strconv"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Additional data describing the firmware setting
+type SettingSchema struct {
+
+	// The type of setting.
+	// +kubebuilder:validation:Enum=Enumeration;String;Integer;Boolean;Password
+	AttributeType string `json:"attribute_type,omitempty"`
+
+	// The allowable value for an Enumeration type setting.
+	AllowableValues []string `json:"allowable_values,omitempty"`
+
+	// The lowest value for an Integer type setting.
+	LowerBound *int `json:"lower_bound,omitempty"`
+
+	// The highest value for an Integer type setting.
+	UpperBound *int `json:"upper_bound,omitempty"`
+
+	// Minimum length for a String type setting.
+	MinLength *int `json:"min_length,omitempty"`
+
+	// Maximum length for a String type setting.
+	MaxLength *int `json:"max_length,omitempty"`
+
+	// Whether or not this setting is read only.
+	ReadOnly *bool `json:"read_only,omitempty"`
+
+	// Whether or not a reset is required after changing this setting.
+	ResetRequired *bool `json:"reset_required,omitempty"`
+
+	// Whether or not this setting's value is unique to this node, e.g.
+	// a serial number.
+	Unique *bool `json:"unique,omitempty"`
+}
+
+// FirmwareSchemaSpec defines the desired state of FirmwareSchema
+type FirmwareSchemaSpec struct {
+	// The schema will be saved in the status section, spec is empty
+}
+
+// FirmwareSchemaStatus defines the observed state of FirmwareSchema
+type FirmwareSchemaStatus struct {
+
+	// The count of the number of times this schema is used by
+	// a HostFirmwareSettings resource
+	ReferenceCount int `json:"referenceCount" required:"true"`
```

Comment:
Sure can move this to Spec.

Reply:
I've been reading up on [Garbage Collection](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/) in k8s, and I think we should be able to use the owner references to manage this automatically (saves a lot of work implementing this in the controller). You _can_ have multiple owner references, which I didn't know before. (Only one of these can be marked as a 'controller', which is what we do for Secrets so that's what confused me. However, we don't [need](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/controller-ref.md) to do that here.)

So it's safe to ditch this field.